### PR TITLE
Books and Works page = v2

### PR DESCRIPTION
**Pull request by jdlrobson**
I'm well and truly stuck.
The remaining thing to make this page mobile friendly+v2 and I can't for the love of me work it out.

Someone please help me find a sane way to pass v2 to the page. All my work in this area is stalled until I can find a safe/sane way to do this :(

```
diff --git a/infogami/utils/view.py b/infogami/utils/view.py
index 45511b0..52102d7 100644
--- a/infogami/utils/view.py
+++ b/infogami/utils/view.py
@@ -161,6 +161,8 @@ def set_error(msg):
     context.error += '\n' + msg
 
 def render_site(url, page):
+    # HACK!!!!
+    page.v2 = True
     return render.site(page)
 
 @public

```

### CODE REVIEW

**Comment by mekarpeles** [comment]
I'm trying my best -- working on this right now!

**Comment by mekarpeles** [comment]
The closest I see from being able to manipulate Types retroactively from outside of infogami is... in `openlibrary/plugins/openlibrary/code.py` after
```
from openlibrary.core import models
models.register_models()
models.register_types()
```
I added:
```
print('=' * 100)
print(dir(client))
print(client._thing_class_registry)
raise
```
And this showed me the following data structure:
```
{'/type/type': <class infogami.infobase.client.Type at 0x7f530db043f8>, '/type/user': <class openlibrary.core.models.User at 0x7f530b07c258>, '/type/edition': <class openlibrary.core.models.Edition at 0x7f530b07c120>, '/type/work': <class openlibrary.core.models.Work at 0x7f530b07c188>, '/type/list': <class openlibrary.core.models.List at 0x7f530b07c2c0>, '/type/library': <class openlibrary.core.models.Library at 0x7f530b07c328>, None: <class openlibrary.core.models.Thing at 0x7f530b07c0b8>, '/type/author': <class openlibrary.core.models.Author at 0x7f530b07c1f0>}
```
Which looks like our types! The question is how Infogami maps these Types to `render_site`

@anandology do you know the answer?

**Comment by mekarpeles** [comment]
Ok, I have an answer, it's just not a fantastic one.

```
diff --git a/openlibrary/core/processors/readableurls.py b/openlibrary/core/processors/readableurls.py
index d4d1e5f..34eb58b 100644
--- a/openlibrary/core/processors/readableurls.py
+++ b/openlibrary/core/processors/readableurls.py
@@ -42,7 +42,12 @@ class ReadableUrlProcessor:
         web.ctx.readable_path = readable_path
         web.ctx.path = real_path
         web.ctx.fullpath = web.ctx.path + web.ctx.query
-        return handler()
+
+        out = handler()
+        V2_TYPES = ['works']
+        if any(web.ctx.path.startswith('/%s/' % _type) for _type in V2_TYPES):
+            out.v2 = True
+        return out
```

We can use `Processor`s which are added in `openlibrary/plugins/openlibrary/code.py` and written either in `openlibrary/plugins/openlibrary/processors.py` or `openlibrary/core/processors` to add checks for pages.

I think (it's a hack but) a long-term choice may be for us to add a new Processor called `V2TypeProcessor` to which would allow us to `openlibrary/plugins/openlibrary/processors.py` which checks the page we're on to see if it matches a `Type` (e.g. `/type/work`) and then add `v2 = True` to the handler output.

In this case, I've extended the `ReadableUrlProcessor` whose job is already to match common `Type` urls (and redirect to the canonical url w/ the e.g. work title slug / author name if/as necessary).

**Comment by jdlrobson** [comment]
Thanks @mekarpeles . It's not perfect but it will have to do (and this is only temporary!)
#1306 can be safely merged before this one and then we can have some mobile friendly versions of these pages.

**Comment by jdlrobson** [comment]
Thanks! #1307 is a minor  thing but also would be good to do before testing/reviewing this branch.

**Comment by mekarpeles** [comment]
Will review today

**Comment by jdlrobson** [comment]
Ready for review!

**Comment by mekarpeles** [comment]
loaded v2 changes on https://dev.openlibrary.org -- noticing:

1) the "Read" buttons on the homepage for the "Classic Books" carousel seem to have disappeared (not sure if it was this or a previous commit)
2) bullets seemingly disappeared from `.book-locate-options`
3) It would be great if the font-size of `class="object"` within `<table class="meta">` were ~`14.4px`

**Comment by jdlrobson** [comment]
Ive not actually changed any styles in this change... so im guessing these problems are also on master?

**Comment by jdlrobson** [comment]
I guess i have changed the html structure though... so maybe some overly specific css rules are no longer applying. A longer shot but possible..

Is this patch still on dev.openlibrary? I am seeing the buttons on the classic books carousel there and it would be good to see those issues on the dev instance so we have a shared reference point!

File: openlibrary/templates/books/custom_carousel.html [hunk 1]
```diff
@@ -1,35 +1,73 @@
 $def with(books=None, title="", url="", key="", min_books=7, test=False)
 
+$def render_carousel_cover(book, lazy):
```
Comment (mekarpeles) [inline openlibrary/templates/books/custom_carousel.html]: Instead of a method here, could this be a `macro`? `CarouselCover`?

**Comment by mekarpeles** [comment]
In the interest of moving a quickly and because additional consolidation will follow, @jdlrobson presents the case that he'd rather not move the `$def` logic to a macro.

I'm alright with that, given the suggestion was evaluated. Moving forward, I'd love to see less python code ($def functions and $jsdef functions) in the template where possible.

One could imagine react as a better supported alternative to webpy's template rendering logic (and I think this is a better long term choice)

**Pull request merged**